<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<HTML>
<HEAD>
<TITLE>phylip</TITLE>
<META NAME="description" CONTENT="phylip">
<META NAME="keywords" CONTENT="phylip">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
</HEAD>
<BODY BGCOLOR="#ccffff">
<P>
<DIV ALIGN=RIGHT>
v3.696
</DIV>
<P>
<DIV ALIGN=CENTER>
<H1>PHYLIP programs and documentation</H1>
</DIV>
<P>
PHYLIP, the PHYLogeny Inference Package, consists of 35 programs.  There are
documentation files for each program, in the form of web pages in HTML 3.2.
There are also documentation web pages for each group of programs, and a
main documentation file that is the basic introduction to the package.
Before running any of the programs you should <A HREF="doc/main.html">read it</A>.
<P>
Below you will find a list of the programs and the documentation files.
The names of the documentation files are highlighted as links that will
take you to those documentation files.
<P>
<DIV ALIGN=CENTER>
<H2>Introduction to PHYLIP</H2>
<P>
<TABLE BORDER=5>
<TR><TD><A HREF="doc/main.html"><FONT SIZE=+1>main documentation file</FONT></A></TD></TR>
</TABLE>
<P>
<H2>Molecular sequence methods</H2>
<P>
<TABLE BORDER=5>
<TR><TD></TD><TD><A HREF="doc/sequence.html"><FONT SIZE=+1>molecular sequence programs documentation file</FONT></A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>protpars</TT></B></FONT></TD><TD><A HREF="doc/protpars.html">protein parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnapars</TT></B></FONT></TD><TD><A HREF="doc/dnapars.html">DNA sequence parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnapenny</TT></B></FONT></TD><TD><A HREF="doc/dnapenny.html">DNA parsimony branch and bound documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnamove</TT></B></FONT></TD><TD><A HREF="doc/dnamove.html">interactive DNA parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnacomp</TT></B></FONT></TD><TD><A HREF="doc/dnacomp.html">DNA compatibility documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnaml</TT></B></FONT></TD><TD><A HREF="doc/dnaml.html">DNA maximum likelihood documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnamlk</TT></B></FONT></TD><TD><A HREF="doc/dnamlk.html">DNA maximum likelihood with clock documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>proml</TT></B></FONT></TD><TD><A HREF="doc/proml.html">Protein sequence maximum likelihood documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>promlk</TT></B></FONT></TD><TD><A HREF="doc/promlk.html">Protein sequence maximum likelihood with clock documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnainvar</TT></B></FONT></TD><TD><A HREF="doc/dnainvar.html">DNA invariants documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dnadist</TT></B></FONT></TD><TD><A HREF="doc/dnadist.html">DNA distance documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>protdist</TT></B></FONT></TD><TD><A HREF="doc/protdist.html">Protein sequence distance documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>restdist</TT></B></FONT></TD><TD><A HREF="doc/restdist.html">Restriction sites and fragments distances documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>restml</TT></B></FONT></TD><TD><A HREF="doc/restml.html">Restriction sites maximum likelihood documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>seqboot</TT></B></FONT></TD><TD><A HREF="doc/seqboot.html">Bootstrapping/Jackknifing documentation file</A></TD></TR>
</TABLE>
<P>
<H2>Distance matrix methods</H2>
<P>
<TABLE BORDER=5>
<TR><TD></TD><TD><A HREF="doc/distance.html"><FONT SIZE=+1>Distance matrix programs documentation file</FONT></A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>fitch</TT></B></FONT></TD><TD><A HREF="doc/fitch.html">Fitch-Margoliash distance matrix method documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>kitsch</TT></B></FONT></TD><TD><A HREF="doc/kitsch.html">Fitch-Margoliash distance matrix with clock documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>neighbor</TT></B></FONT></TD><TD><A HREF="doc/neighbor.html">Neighbor-Joining and UPGMA method documentation file</A></TD></TR>
</TABLE>
<P>
<H2>Gene frequencies and continuous characters</H2>
<P>
<TABLE BORDER=5>
<TR><TD></TD><TD><A HREF="doc/contchar.html"><FONT SIZE=+1>Continuous characters and gene frequencies documentation file</FONT></A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>contml</TT></B></FONT></TD><TD><A HREF="doc/contml.html">Maximum likelihood continuous characters and gene frequencies documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>contrast</TT></B></FONT></TD><TD><A HREF="doc/contrast.html">Contrast method documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>gendist</TT></B></FONT></TD><TD><A HREF="doc/gendist.html">Genetic distance documentation file</A></TD></TR>
</TABLE>
<P>
<H2>Discrete characters methods</H2>
<P>
<TABLE BORDER=5>
<TR><TD></TD><TD><A HREF="doc/discrete.html"><FONT SIZE=+1>Discrete characters methods documentation file</FONT></A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>pars</TT></B></FONT></TD><TD><A HREF="doc/pars.html">Unordered multistate parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>mix</TT></B></FONT></TD><TD><A HREF="doc/mix.html">Mixed method parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>penny</TT></B></FONT></TD><TD><A HREF="doc/penny.html">Branch and bound mixed method parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>move</TT></B></FONT></TD><TD><A HREF="doc/move.html">Interactive mixed method parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dollop</TT></B></FONT></TD><TD><A HREF="doc/dollop.html">Dollo and polymorphism parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dolpenny</TT></B></FONT></TD><TD><A HREF="doc/dolpenny.html">Dollo and polymorphism branch and bound parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>dolmove</TT></B></FONT></TD><TD><A HREF="doc/dolmove.html">Dollo and polymorphism interactive parsimony documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>clique</TT></B></FONT></TD><TD><A HREF="doc/clique.html">0/1 characters compatibility method documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>factor</TT></B></FONT></TD><TD><A HREF="doc/factor.html">Character recoding program documentation file</A></TD></TR>
</TABLE>
<P>
<H2>Tree drawing, consensus, tree editing, tree distances</H2>
<P>
<TABLE BORDER=5>
<TR><TD></TD><TD><A HREF="doc/draw.html">Tree drawing programs documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>drawgram</TT></B></FONT></TD><TD><A HREF="doc/drawgram.html">Rooted tree drawing program documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>drawtree</TT></B></FONT></TD><TD><A HREF="doc/drawtree.html">Unrooted tree drawing program documentation file</A></TD></TR>
<TR><TD>&nbsp;</TD><TD>&nbsp;</TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>consense</TT></B></FONT></TD><TD><A HREF="doc/consense.html">Consensus tree program documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>treedist</TT></B></FONT></TD><TD><A HREF="doc/treedist.html">Tree distance program documentation file</A></TD></TR>
<TR><TD><FONT SIZE=+1><B><TT>retree</TT></B></FONT></TD><TD><A HREF="doc/retree.html">interactive tree rearrangement program documentation file</A></TD></TR>
</TABLE>
</DIV>
</BODY>
</HTML>
